src/Facility.CodeGen.CSharp/CSharpGenerator.cs

```diff
@@ -356,11 +391,108 @@ private CodeGenFile GenerateDto(ServiceDtoInfo dtoInfo, Context context)
 									}
 								}
 
-								if (validatableFields.Count != 0)
+								if (validateFields.Count != 0)
+								{
+									code.WriteLineSkipOnce();
+									foreach (var fieldInfo in validateFields)
+									{
+										var type = context.GetFieldType(fieldInfo);
+										var propertyName = context.GetFieldPropertyName(fieldInfo);
+										var validation = fieldInfo.Validation!;
+
+										switch (type.Kind)
+										{
+											case ServiceTypeKind.Enum:
+											{
+												code.WriteLine($"if ({propertyName} != null && !{propertyName}.Value.IsDefined())");
+												using (code.Indent())
+													code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Must be an expected enum value.\");");
+
+												break;
+											}
+
+											case ServiceTypeKind.String:
+											{
+												var validRange = validation.LengthRange;
+												if (validRange != null)
+												{
+													if (validRange.Minimum != null)
+													{
+														code.WriteLine($"if ({propertyName} != null && {propertyName}.Length < {validRange.Minimum})");
+														using (code.Indent())
+															code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Length must be at least {validRange.Minimum}.\");");
+													}
+
+													if (validRange.Maximum != null)
+													{
+														code.WriteLine($"if ({propertyName} != null && {propertyName}.Length > {validRange.Maximum})");
+														using (code.Indent())
+															code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", \"Length must be at most {validRange.Maximum}.\");");
+													}
+												}
+
+												var validPattern = validation.RegexPattern;
+												if (validPattern != null)
+												{
+													var regexField = $"s_valid{propertyName}Regex";
+													code.WriteLine($"if ({propertyName} != null && !{regexField}.IsMatch({propertyName}))");
+													using (code.Indent())
+														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", $\"Must match regular expression: {{{regexField}}}.\");");
```

Comment:
```suggestion
														code.WriteLine($"return ServiceDataUtility.GetInvalidFieldErrorMessage(\"{fieldInfo.Name}\", $\"Must match regular expression: {{{regexField}}}\");");
```

Nit: I omit the trailing period when I use the colon pattern, which I use when I want to display a string without worrying about delimiters.